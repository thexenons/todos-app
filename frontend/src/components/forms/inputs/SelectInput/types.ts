import { FieldValues, RegisterOptions } from "react-hook-form";

import { InputBaseProps } from "../InputBase/types";

export interface SelectInputChoice {
	value: InputBaseProps["value"];
	label: string;
}

export interface SelectInputProps extends InputBaseProps {
	choices: SelectInputChoice[];
	name: string;
	rules?: Omit<
		RegisterOptions<FieldValues, string>,
		"valueAsNumber" | "valueAsDate" | "setValueAs" | "disabled"
	>;
}
