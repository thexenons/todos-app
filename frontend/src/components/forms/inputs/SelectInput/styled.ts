import styled, { css } from "styled-components";

import FloatingPanel from "../../../layout/FloatingPanel";
import { FloatingPanelContent } from "../../../layout/FloatingPanel/styled";

export const SelectFloatingPanel = styled(FloatingPanel)`
	${FloatingPanelContent} {
		${({ theme }) => theme.mixins.padding({ all: 0 })}
	}
`;

export const OptionsList = styled.ul`
	list-style: none;
	${({ theme }) => theme.mixins.margin({ all: 0 })};
	${({ theme }) => theme.mixins.padding({ x: 0, y: theme.spacing(1) })};
`;

export const OptionItem = styled.li<{
	$isActive?: boolean;
	$isCurrent?: boolean;
}>`
	cursor: pointer;
	${({ theme }) =>
		theme.mixins.padding({ y: theme.spacing(0.5), x: theme.spacing(1) })}

	${({ $isCurrent }) =>
		$isCurrent &&
		css`
			font-weight: 900;
		`}

	${({ $isActive }) =>
		$isActive &&
		css`
			background: rgba(255, 255, 255, 0.5);
		`}
`;

export const EmptyChoices = styled.div`
	${({ theme }) =>
		theme.mixins.padding({ y: theme.spacing(0.5), x: theme.spacing(1) })}
`;
