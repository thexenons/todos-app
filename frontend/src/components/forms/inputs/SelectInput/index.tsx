import { forwardRef } from "react";

import InputBase from "../InputBase";
import useSelectInput from "./hooks/useSelectInput";
import * as S from "./styled";
import { SelectInputProps } from "./types";

const SelectInput = forwardRef<HTMLInputElement, SelectInputProps>(
	(props, ref) => {
		const { choices, inputProps, floatingPanelProps, optionItem } =
			useSelectInput(props, ref);

		return (
			<S.SelectFloatingPanel
				{...floatingPanelProps}
				content={
					<S.OptionsList>
						{choices.map((choice, index) => (
							<S.OptionItem
								key={String(choice.value)}
								{...optionItem.getProps(choice, index)}
							/>
						))}
						{!choices.length && <S.EmptyChoices>Empty choices</S.EmptyChoices>}
					</S.OptionsList>
				}
			>
				<InputBase {...inputProps} />
			</S.SelectFloatingPanel>
		);
	}
);

SelectInput.displayName = "SelectInput";

export default SelectInput;
