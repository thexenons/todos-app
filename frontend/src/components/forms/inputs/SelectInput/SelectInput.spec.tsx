import { render } from "@testing-library/react";

import InputTestComponent from "../../../../tests/InputTestComponent";
import SelectInput from ".";

const choices = [
	{
		value: 1,
		label: "Opción 1",
	},
	{
		value: 2,
		label: "Opción 2",
	},
];

describe("SelectInput", () => {
	it("Render", () => {
		const { container } = render(
			<InputTestComponent>
				<SelectInput name="test" choices={choices} />
			</InputTestComponent>
		);
		expect(container).toMatchSnapshot();
	});

	it("should set ref", () => {
		let inputElement: HTMLInputElement | undefined;
		const handleRef = (ref: HTMLInputElement) => {
			inputElement = ref;
		};
		render(
			<InputTestComponent>
				<SelectInput name="test" ref={handleRef} choices={choices} />
			</InputTestComponent>
		);
		expect(inputElement).not.toBeNull();
	});
});
