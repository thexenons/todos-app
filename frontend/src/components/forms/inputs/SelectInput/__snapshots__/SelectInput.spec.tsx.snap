// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SelectInput > Render 1`] = `
<div>
  <form>
    <div
      role="button"
      tabindex="0"
    >
      <div
        class="styled__InputWrapper-sc-8zf6s0-2 iKmMVX"
        data-has-error="false"
      >
        <input
          class="styled__InputElement-sc-8zf6s0-1 kBsvAW"
          data-has-placeholder="false"
          id=":r0:"
          name="test"
          placeholder=" "
          type="text"
          value=""
        />
        <label
          class="styled__InputLabel-sc-8zf6s0-0 kSURlg"
          for=":r0:"
        />
      </div>
    </div>
  </form>
</div>
`;
